nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java

```diff
@@ -352,95 +355,95 @@ protected void processBatchOfFiles(final List<Path> files, final ProcessContext
 
         for (final Path file : files) {
 
-            ugi.doAs(new PrivilegedAction<Object>() {
-                @Override
-                public Object run() {
-                    FlowFile flowFile = session.create(parentFlowFile);
-                    try {
-                        final String originalFilename = file.getName();
-                        final Path outputDirPath = getNormalizedPath(context, OUTPUT_DIRECTORY, parentFlowFile);
-                        final Path newFile = new Path(outputDirPath, originalFilename);
-                        final boolean destinationExists = hdfs.exists(newFile);
-                        // If destination file already exists, resolve that
-                        // based on processor configuration
-                        if (destinationExists) {
-                            switch (processorConfig.getConflictResolution()) {
-                                case REPLACE_RESOLUTION:
-                                    // Remove destination file (newFile) to replace
-                                    if (hdfs.delete(newFile, false)) {
-                                        getLogger().info("deleted {} in order to replace with the contents of {}",
-                                                new Object[]{newFile, flowFile});
-                                    }
-                                    break;
-                                case IGNORE_RESOLUTION:
-                                    session.transfer(flowFile, REL_SUCCESS);
-                                    getLogger().info(
-                                            "transferring {} to success because file with same name already exists",
-                                            new Object[]{flowFile});
-                                    return null;
-                                case FAIL_RESOLUTION:
-                                    session.transfer(session.penalize(flowFile), REL_FAILURE);
-                                    getLogger().warn(
-                                            "penalizing {} and routing to failure because file with same name already exists",
-                                            new Object[]{flowFile});
-                                    return null;
-                                default:
-                                    break;
-                            }
+            ugi.doAs((PrivilegedAction<Object>) () -> {
```

Comment:
This way ModeHDFS is throwing a NullPointerException when the GSSException occurs because it tries the next file after it did a rollback.

Here's a patch (against the current state of this PR) that would help:
```
Subject: [PATCH] Changes base on code review comments
---
Index: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java b/nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java
--- a/nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java	(revision 024860a2d8dccbaed8e98d280e1afd22dd0b7201)
+++ b/nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/MoveHDFS.java	(date 1712341636424)
@@ -47,9 +47,11 @@
 import org.apache.nifi.processor.exception.ProcessException;
 import org.apache.nifi.processor.util.StandardValidators;
 import org.apache.nifi.util.StopWatch;
+import org.ietf.jgss.GSSException;
 
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.security.PrivilegedAction;
 import java.security.PrivilegedExceptionAction;
 import java.util.ArrayList;
@@ -57,6 +59,7 @@
 import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedBlockingQueue;
@@ -325,16 +328,20 @@
             queueLock.unlock();
         }
 
-        processBatchOfFiles(files, context, session, flowFile);
+        try {
+            processBatchOfFiles(files, context, session, flowFile);
 
+            session.remove(flowFile);
+        } catch (UncheckedIOException e) {
+            handleAuthErrors(e, session, context);
+        }
+
         queueLock.lock();
         try {
             processing.removeAll(files);
         } finally {
             queueLock.unlock();
         }
-
-        session.remove(flowFile);
     }
 
     protected void processBatchOfFiles(final List<Path> files, final ProcessContext context,
@@ -433,15 +440,14 @@
                     final String transitUri = hdfs.getUri() + StringUtils.prependIfMissing(outputPath, "/");
                     session.getProvenanceReporter().send(flowFile, transitUri);
                     session.transfer(flowFile, REL_SUCCESS);
-
                 } catch (final Throwable t) {
-                    if (handleAuthErrors(t, session, context)) {
-                        return null;
-                    } else {
-                        getLogger().error("Failed to rename on HDFS due to {}", new Object[]{t});
-                        session.transfer(session.penalize(flowFile), REL_FAILURE);
-                        context.yield();
-                    }
+                    final Optional<GSSException> causeOptional = findCause(t, GSSException.class, gsse -> GSSException.NO_CRED == gsse.getMajor());
+                    if (causeOptional.isPresent()) {
+                        throw new UncheckedIOException(new IOException(causeOptional.get()));
+                    }
+                    getLogger().error("Failed to rename on HDFS due to {}", new Object[]{t});
+                    session.transfer(session.penalize(flowFile), REL_FAILURE);
+                    context.yield();
                 }
                 return null;
             });

```

Reply:
I think the patch didn't apply correctly. The NullPointerException still occurs. In the processBatchOfFiles method's  catch should look like this:

```java
    catch (final Throwable t) {
                    final Optional<GSSException> causeOptional = findCause(t, GSSException.class, gsse -> GSSException.NO_CRED == gsse.getMajor());
                    if (causeOptional.isPresent()) {
                        throw new UncheckedIOException(new IOException(causeOptional.get()));
                    }
                    getLogger().error("Failed to rename on HDFS due to {}", new Object[]{t});
                    session.transfer(session.penalize(flowFile), REL_FAILURE);
                    context.yield();
                }
```